<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment</b>
    
Hi, it's Sveta again. I was the reviewer on one of your previous projects, and I am going to review this one as well. I apologize for this egregious delay, and I hope it did not cause too much inconvenience. 
    
    
I will be using the same color marking:
    
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
Great solutions and ideas that can and should be used in the future are in green comments.   
</div>    
    
    
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>

Yellow color indicates what should be optimized. This is not necessary, but it will be great if you make changes to this project.
</div>      
    
    
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>

Issues that need to be corrected to get right results are indicated in red comments. Note that the project cannot be accepted until these issues are resolved.
</div>    

<hr>
    
**Please, use some color other than those listed to highlight answers to my comments.**
I would also ask you **not to change, move or delete my comments** so that it would be easier for me to navigate during the next review.
    

    
    
<hr>
    
<font color='dodgerblue'>**A few words about the project:**</font> this project is hard, but you did a pretty good job. There are several issues that need your attention. Please take a look. 
    

</div>

<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment 2</b>


Thank you for the modifications. There's just a couple of questions. Would you take a look? 
    
  
</div>
<hr>
<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment 3</b>


Thank you for the correcctions! There's only one issue this time. Please take a look and reconsider the conclusions if needed.
    
  
</div>
<hr>
<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment 4</b>


You did a great job, thank you so much! 😊 I do not have any questions, so our project has passed code review. Congratulations 😊
    

    
Good luck! 😊   
    
    
    
    
<hr>
    
    
Best regards,
    
Sveta
     
</div>

# Predicting Customer Churn in Beta Bank: A Supervised Learning Approach


This project aims to predict customer churn in Beta Bank by analyzing historical data on client behavior and contract terminations. Using supervised learning techniques, we seek to develop a model to identify customers likely to leave the bank soon. Our goal is to provide actionable insights to Beta Bank to optimize retention strategies and address customer attrition effectively.


In [1]:
import pandas as pd
from sklearn.preprocessing import OrdinalEncoder
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
from sklearn.metrics import f1_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

In [2]:
# 'Tenure' column is the only one with missing values
df = pd.read_csv('/datasets/Churn.csv')
display(df)
print(df.info())

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0        2.0       0.00              1          1               1   
1        1.0   83807.86              1          0               1   
2        8.0  159660.80              3          1               0   
3        1.0       0.00              2          0               0   
4        2.0  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995     5.0       0.00              2          1               0   
9996    10.0   57369.61              1          1               1   
9997     7.0       0.00              1          0               1   
9998     3.0   75075.31              2          1               0   
9999     NaN  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB
None


In [3]:
# Remove a few columns because they're unnecessary
df.drop(columns=['RowNumber', 'CustomerId', 'Surname'], inplace=True)

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment</h2>
    
So, we can already notice that some of the data types can be changed. For instance, `HasCrCard` is probably a boolean type.     
</div>

In [4]:
# Checking for any obvious missing values and duplicated rows
print(df['Tenure'].value_counts())
print()
print(df.duplicated().sum())

1.0     952
2.0     950
8.0     933
3.0     928
5.0     927
7.0     925
4.0     885
9.0     882
6.0     881
10.0    446
0.0     382
Name: Tenure, dtype: int64

0


In [5]:
print(df.describe())

        CreditScore           Age       Tenure        Balance  NumOfProducts  \
count  10000.000000  10000.000000  9091.000000   10000.000000   10000.000000   
mean     650.528800     38.921800     4.997690   76485.889288       1.530200   
std       96.653299     10.487806     2.894723   62397.405202       0.581654   
min      350.000000     18.000000     0.000000       0.000000       1.000000   
25%      584.000000     32.000000     2.000000       0.000000       1.000000   
50%      652.000000     37.000000     5.000000   97198.540000       1.000000   
75%      718.000000     44.000000     7.000000  127644.240000       2.000000   
max      850.000000     92.000000    10.000000  250898.090000       4.000000   

         HasCrCard  IsActiveMember  EstimatedSalary        Exited  
count  10000.00000    10000.000000     10000.000000  10000.000000  
mean       0.70550        0.515100    100090.239881      0.203700  
std        0.45584        0.499797     57510.492818      0.402769  
min    

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
Are there any duplicates in `CustomerId` column? 

</div>
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b>  Reviewer's comment </b>
   

- It may be useful to print some statistics to evaluate the dataset. Simple `describe` method would suffice. 


- We can also convert column names from camel to snake case. 


- Unnecessary columns can be removed.

</div>
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2 </h2>
    
Looks better! 

</div>

In [6]:
# Filling in the missing values in the 'Tenure' column with the mean because there aren't any outliers or anything 
df['Tenure'].fillna(df['Tenure'].median(), inplace=True)

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment</h2>
    
Yes, we should fill in these gaps. 

</div>
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

However, you said we could use the average because "there aren't any outliers or anything". How do you know that? Isn't it safer to just use the median? 
    
</div>

^ I know this because I looked at the value_counts of the column and all the values were mostly equal except for the maximum and the minimum values which I think makes sense. Also, with the .describe stats, we can see that the mean basically is the median so it won't make much of a difference anyway. But I will change it to median

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2 </h2>
    
Makes sense. Thank you for the clarification. 

</div>

In [24]:
# Can use label encoding to replace the Geography and Gender columns since there's only 2-3 different values in each
# Nothing looks out of place to me
print(df['Geography'].value_counts())
print()
print(df['Gender'].value_counts())
print()
print(df['Balance'].value_counts())
print()
print(df['NumOfProducts'].value_counts())
print()
print(df['Age'].value_counts())
print()
print(df['IsActiveMember'].sample(5))

France     5014
Germany    2509
Spain      2477
Name: Geography, dtype: int64

Male      5457
Female    4543
Name: Gender, dtype: int64

0.00         3617
105473.74       2
130170.82       2
72594.00        1
139723.90       1
             ... 
130306.49       1
92895.56        1
132005.77       1
166287.85       1
104001.38       1
Name: Balance, Length: 6382, dtype: int64

1    5084
2    4590
3     266
4      60
Name: NumOfProducts, dtype: int64

37    478
38    477
35    474
36    456
34    447
     ... 
92      2
88      1
82      1
85      1
83      1
Name: Age, Length: 70, dtype: int64

9565    1
5878    1
3265    0
4111    1
1474    1
Name: IsActiveMember, dtype: int64


<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
Let's analyze other columns to make sure the data looks normal. 
    
</div>
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 2</b>
    
There are more than 3 columns :)     
</div>
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2>  Reviewer's comment 3 </h2>
   

When you have a lot of values, use charts instead of `value_counts()`, since the latter will not give you enough information. 
</div>

In [8]:
# OHE for Geography and Gender

df_ohe = pd.get_dummies(df, columns=['Geography', 'Gender'], drop_first=True) # Fixed

display(df_ohe)

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0             619   42     2.0       0.00              1          1   
1             608   41     1.0   83807.86              1          0   
2             502   42     8.0  159660.80              3          1   
3             699   39     1.0       0.00              2          0   
4             850   43     2.0  125510.82              1          1   
...           ...  ...     ...        ...            ...        ...   
9995          771   39     5.0       0.00              2          1   
9996          516   35    10.0   57369.61              1          1   
9997          709   36     7.0       0.00              1          0   
9998          772   42     3.0   75075.31              2          1   
9999          792   28     5.0  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Exited  Geography_Germany  \
0                  1        101348.88       1                  0   
1                  1        112542.58       0                  0   
2                  0        113931.57       1                  0   
3                  0         93826.63       0                  0   
4                  1         79084.10       0                  0   
...              ...              ...     ...                ...   
9995               0         96270.64       0                  0   
9996               1        101699.77       0                  0   
9997               1         42085.58       1                  0   
9998               0         92888.52       1                  1   
9999               0         38190.78       0                  0   

      Geography_Spain  Gender_Male  
0                   0            0  
1                   1            0  
2                   0            0  
3                   0            0  
4                   1            0  
...               ...          ...  
9995                0            1  
9996                0            1  
9997                0            0  
9998                0            1  
9999                0            0  

[10000 rows x 12 columns]

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 3</b>
    
Please don't forget about `drop_first` parameter.     
</div>
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2>   Reviewer's comment 4 ✔️</h2>
    
Good. Using `drop_first=True` may help you avoid multicollinearity.


</div>


<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2 </h2>
    
> I used label encoding because I was told that OHE should generally only be used for regression models
    
    
Like many other methods, OHE methods have pros and cons, but it does not mean we should use them for regression only. OHE works fine with classification, but it increases the dimension.
    
    
> But when I tried to train the model, it gave me an error: 'ValueError: could not convert string to float'
    
I do not see the code, so I cannot suggest anything here. 
   
    
    
> I thought that I might just remove the surname column, but I think it could have some importance in the data?
    
    
If a feature does not influence people's decisions regarding leaving the bank, then we do not need to include these columns in the analysis. Row numbers, ids, and last names should not have such influence, so it is save to drop them. It will reduce the size of the dataframe and make the calculations faster.  
    
    
    
</div><div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 2</b>
    
 
    
> If you still think I should change it to OHE, that's no problem, I'm just explaining why I did it this way
    
    
It is not a probelm that you use ordinal encoding. The problem is in the way you are using it. We must encode categorical data only. 
</div>

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
- Only categorical features should be encoded.
   

- Please specify why you decided to choose ordinal encoding. The OHE methods seem more preferable here. For instance, one-hot encoding helps us ensure the model does not misinterpret categorical features as having some ordinal relationships.


</div>
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2>  Reviewer's comment </h2>
   

[PEP8](https://peps.python.org/pep-0008/) states that one should always put import at the top of the file. It's a good practice, since everyone who is going to read the project, can immediately figure out what modules need to be installed. 

</div>

^ I usually would haha but I've been told before that there's no reason to put them all at the top and got flagged for it, so I've been spreading them out. I'll change that

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
I gave you a reason in the comment above :) 

</div>

In [9]:
# initialize temp train df and test df
# I split 20% because it is a large dataset and I'm splitting it twice which would only leave 50% for training if I used 25%

df_train_temp, df_test = train_test_split(
    df_ohe, 
    test_size=.2, # Split 20% of data to make test set
    random_state=12345)

# initialize valid df and actual train df
df_train, df_valid = train_test_split(
    df_train_temp, 
    test_size=0.2,  # Split another 20% of data to make validation set
    random_state=54321)

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2>  Reviewer's comment </h2>
   

There's a `stratify` parameter that help us ensure the random split saves the initial proportion of values in some column.   
    
</div>

In [10]:
# Initialize variables
features = df_ohe.drop(['Exited'], axis=1)
target = df_ohe['Exited']

features_train = df_train.drop(['Exited'], axis=1)
target_train = df_train['Exited']

features_test = df_test.drop(['Exited'], axis=1)
target_test = df_test['Exited']

features_valid = df_valid.drop(['Exited'], axis=1)
target_valid = df_valid['Exited']

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment </h2>
    
Correct.
    
</div>
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2>  Reviewer's comment </h2>
   

After you successfully split the data, you can scale it, it never hurts.

</div>

In [11]:
# Initialize Decision Tree model and train it without class_weight

'''
best_f1 = 0
best_accuracy = 0
best_est = 0
for est in range(1, 21):
    tree1 = DecisionTreeClassifier(random_state=12345, max_depth=est)
    tree1.fit(features_train, target_train)
    # Predict on validation set to calculate F1 score
    tree1_predicted_valid = tree1.predict(features_valid)
    # Calculate accuracy score on validation set
    accuracy = accuracy_score(target_valid, tree1_predicted_valid)
    # Calculate F1 score
    f1 = f1_score(target_valid, tree1_predicted_valid)
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_est = est
    if f1 > best_f1:
        best_f1 = f1
        best_est = est

print("Accuracy of the best model on the validation set (max_depth = {}): {}".format(best_est, best_accuracy)) # Best for both
print("Best f1 score of the best model on the validation set (max_depth = {}): {}".format(best_est, best_f1)) # is max_depth=6
'''
tree1 = DecisionTreeClassifier(random_state=12345, max_depth=6)
tree1.fit(features_train, target_train)

tree1_predicted_valid = tree1.predict(features_valid)

print()
print('Without class weight:')
print()
print('Recall:', recall_score(target_valid, tree1_predicted_valid))
print()
print('Precision:', precision_score(target_valid, tree1_predicted_valid))
print()
print('F1 score:', f1_score(target_valid, tree1_predicted_valid))
print()
print('Valid acc:', accuracy_score(target_valid, tree1_predicted_valid))
print()
print('AUC-ROC value:', roc_auc_score(target_valid, tree1_predicted_valid))


Without class weight:

Recall: 0.44983818770226536

Precision: 0.7679558011049724

F1 score: 0.5673469387755101

Valid acc: 0.8675

AUC-ROC value: 0.7086526337426897


In [12]:
# Initialize Decision Tree model and train it with class_weight

'''
best_f1 = 0
best_accuracy = 0
best_est = 0
for est in range(1, 21):
    tree2 = DecisionTreeClassifier(random_state=12345, max_depth=est, class_weight='balanced')
    tree2.fit(features_train, target_train)
    # Predict on validation set to calculate F1 score
    tree2_predicted_valid = tree2.predict(features_valid)
    # Calculate accuracy score on validation set
    accuracy = accuracy_score(target_valid, tree2_predicted_valid)
    # Calculate F1 score
    f1 = f1_score(target_valid, tree2_predicted_valid)
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_est = est
    if f1 > best_f1:
        best_f1 = f1
        best_est = est

print("Accuracy of the best model on the validation set (max_depth = {}): {}".format(best_est, best_accuracy)) # Best for both
print("Best f1 score of the best model on the validation set (max_depth = {}): {}".format(best_est, best_f1)) # is max_depth=16
'''
tree2 = DecisionTreeClassifier(random_state=12345, max_depth=16, class_weight='balanced')
tree2.fit(features_train, target_train)

tree2_predicted_valid = tree2.predict(features_valid)

print()
print('With class weight:')
print()
print('Recall:', recall_score(target_valid, tree2_predicted_valid))
print()
print('Precision:', precision_score(target_valid, tree2_predicted_valid))
print()
print('F1 score:', f1_score(target_valid, tree2_predicted_valid))
print()
print('Valid acc:', accuracy_score(target_valid, tree2_predicted_valid))
print()
print('AUC-ROC value:', roc_auc_score(target_valid, tree2_predicted_valid))


With class weight:

Recall: 0.5307443365695793

Precision: 0.49101796407185627

F1 score: 0.5101088646967341

Valid acc: 0.803125

AUC-ROC value: 0.69953173451252


<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 2 </b>
    
Please add the code if you write about its results in the conclusions.
</div>
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 3 </h2>
    
Good! The results may change though. 
    
</div>

**Unbalanced DecisionTreeClassifier class_weight Conclusion**

A balanced class_weight in the DecisionTree model lowered the accuracy by ~.06, raised the recall by ~.11, lowered f1 score by ~.05, lowered precision by ~.27, and left the AUC-ROC basically the same so I think the one without class_weight is better.

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

What can be concluded from the output above? 
    
</div>

^ I moved the balance check to after the conclusion for the three models

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Here and further, let's add AUC-ROC score. 
    
</div>

Done

In [13]:
# Try a RandomForestClassifier without balance

'''
# Test to find optimal n_estimators value
best_f1 = 0
best_accuracy = 0
best_est = 0
for est in range(1, 101): #
    forest1 = RandomForestClassifier(random_state=54321, n_estimators=est) # set number of n_estimators
    forest1.fit(features_train, target_train) # train model on training set
    forest1_predicted_valid = forest1.predict(features_valid)
    accuracy = accuracy_score(target_valid, forest1_predicted_valid) # calculate accuracy score on validation set
    f1 = f1_score(target_valid, forest1_predicted_valid)
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_est = est
    if f1 > best_f1:
        best_f1 = f1
        best_est = est

print("Accuracy of the best model on the validation set (n_estimators = {}): {}".format(best_est, best_accuracy)) # 96
print("Best f1 score of the best model on the validation set (n_estimators = {}): {}".format(best_est, best_f1)) # 96
'''

forest1 = RandomForestClassifier(random_state=54321, n_estimators=96)
forest1.fit(features_train, target_train)

forest1_predicted_valid = forest1.predict(features_valid)

print('Without class_weight:')
print()
print('Recall:', recall_score(target_valid, forest1_predicted_valid))
print()
print('Precision:', precision_score(target_valid, forest1_predicted_valid))
print()
print('F1 score:', f1_score(target_valid, forest1_predicted_valid))
print()
print('Valid acc:', accuracy_score(target_valid, forest1_predicted_valid))
print()
print('AUC-ROC value:', roc_auc_score(target_valid, forest1_predicted_valid))


Without class_weight:

Recall: 0.47896440129449835

Precision: 0.7956989247311828

F1 score: 0.5979797979797981

Valid acc: 0.875625

AUC-ROC value: 0.7247649272158008


In [14]:
# Try a RandomForestClassifier with balance

'''
# Test to find optimal n_estimators value
best_f1 = 0
best_accuracy = 0
best_est = 0
for est in range(1, 101): #
    forest2 = RandomForestClassifier(random_state=54321, n_estimators=est, class_weight='balanced') # set number of n_estimators
    forest2.fit(features_train, target_train) # train model on training set
    forest2_predicted_valid = forest2.predict(features_valid)
    accuracy = accuracy_score(target_valid, forest2_predicted_valid) # calculate accuracy score on validation set
    f1 = f1_score(target_valid, forest2_predicted_valid)
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_est = est
    if f1 > best_f1:
        best_f1 = f1
        best_est = est

print("Accuracy of the best model on the validation set (n_estimators = {}): {}".format(best_est, best_accuracy)) # 39
print("Best f1 score of the best model on the validation set (n_estimators = {}): {}".format(best_est, best_f1)) # 39
'''

forest2 = RandomForestClassifier(random_state=54321, n_estimators=39, class_weight='balanced')
forest2.fit(features_train, target_train)

forest2_predicted_valid = forest2.predict(features_valid)

print('With class_weight:')
print()
print('Recall:', recall_score(target_valid, forest2_predicted_valid))
print()
print('Precision:', precision_score(target_valid, forest2_predicted_valid))
print()
print('F1 score:', f1_score(target_valid, forest2_predicted_valid))
print()
print('Valid acc:', accuracy_score(target_valid, forest2_predicted_valid))
print()
print('AUC-ROC value:', roc_auc_score(target_valid, forest2_predicted_valid))

With class_weight:

Recall: 0.4627831715210356

Precision: 0.7606382978723404

F1 score: 0.5754527162977867

Valid acc: 0.868125

AUC-ROC value: 0.7139632356443287


**Unbalanced RandomForestClassifier class_weight Conclusion**

A balanced class weight reduced recall by ~.01, lowered precision by ~.03, lowered f1 score by ~.02, lowered accuracy by ~.01, and lowered AOC-ROC by ~.01 so I think the model without a balanced class weight performs better.

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 2</b>
    

Add the code without class_weight here.
</div>

In [15]:
# Trying out a logistic regression model without class_weight variable

logReg1 = LogisticRegression(solver='liblinear', random_state=12345)
logReg1.fit(features_train, target_train)

log1_predicted_valid = logReg1.predict(features_valid)

print('Without class_weight:')
print()
print('Recall:', recall_score(target_valid, log1_predicted_valid))
print()
print('Precision:', precision_score(target_valid, log1_predicted_valid))
print()
print('F1 score:', f1_score(target_valid, log1_predicted_valid))
print()
print('Valid acc:', accuracy_score(target_valid, log1_predicted_valid))
print()
print('AUC-ROC value:', roc_auc_score(target_valid, log1_predicted_valid))

Without class_weight:

Recall: 0.009708737864077669

Precision: 0.6

F1 score: 0.01910828025477707

Valid acc: 0.8075

AUC-ROC value: 0.5040797755935416


In [16]:
# Trying out a logistic regression model with class_weight variable

logReg2 = LogisticRegression(solver='liblinear', random_state=12345, class_weight='balanced')
logReg2.fit(features_train, target_train)

log2_predicted_valid = logReg2.predict(features_valid)

print('With class_weight:')
print()
print('Recall:', recall_score(target_valid, log2_predicted_valid))
print()
print('Precision:', precision_score(target_valid, log2_predicted_valid))
print()
print('F1 score:', f1_score(target_valid, log2_predicted_valid))
print()
print('Valid acc:', accuracy_score(target_valid, log2_predicted_valid))
print()
print('AUC-ROC value:', roc_auc_score(target_valid, log2_predicted_valid))

With class_weight:

Recall: 0.6731391585760518

Precision: 0.33015873015873015

F1 score: 0.44302449414270506

Valid acc: 0.673125

AUC-ROC value: 0.6731303848650979


**Unbalanced LogisticRegression class_weight Conclusion**

Balancing the class weight for the LogisticRegression model improved the recall by ~.66, reduced the precision by ~.27, improved the f1 score by ~.42, reduced the accuracy by ~.13, and increased the AOC-ROC by ~.17. I think class_weight does help this model but it doesn't really matter because it's still really bad.

I think in general, class_weight='balanced' made the models worse, but not by that much.

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Code is missing here as well. We have to show the result, not just tell about it. 
    
</div>
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 3 </h2>
    
Indeed, it looks like it did not help. However, we still need to check everything again, since the results may change. 
    
</div>

# Model Stats Comparisons Without Class_Weight='balanced'

**Recall:**
* Worst - LogisticRegression
* Best - RandomForestClassifier

**Precision:**
* Worst - LogisticRegression
* Best - RandomForestClassifier

**F1 Score:**
* Worst - LogisticRegression
* Best - RandomForestClassifier

**Valid Accuracy:**
* Worst - LogisticRegression
* Best - RandomForestClassifier

**AUC-ROC Value:**
* Worst - LogisticRegression
* Best - RandomForestClassifier

It's pretty clear that LogisticRegression is the worst and RandomForest is the best, at least before balancing the classes

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

- Please add the code you used here. 
    

- Try other models as well.


- Compare the results. 

</div>
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 2</b>
    

Yes, we should provide the code. Conclusions are not reliable without the results, right? 
    

</div>

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b>  Reviewer's comment </b>
   

It may be a bit confusing that a conclusion is defined with the level-1 heading :) 
</div>

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
Before we move on to the resampling, let's check whether a simple `class_weight` parameter can help here. 
    
</div>

In [17]:
# Check the balance of classes

print(df_ohe['Exited'].value_counts())
print(df_train['Exited'].value_counts())

0    7963
1    2037
Name: Exited, dtype: int64
0    5099
1    1301
Name: Exited, dtype: int64


There is an imbalance in the classes. 1.0 has about 1/4 the amount of 0.0.

In 'real' terms, it means that about a quarter of the customers have already left the bank.

In [18]:
# Downsample + Upsample
# My plan is to upsample by 2 and downsample by 2 since that would make the two classes a very close number. ~4000
# I wanted to do this so that I create and delete an equal number so it's even more? balanced?
# It just sounded like a good idea

# Separate majority and minority classes
df_majority = df_train[df_train['Exited'] == 0]
df_minority = df_train[df_train['Exited'] == 1]

# Downsample majority class
df_majority_downsampled = df_majority.sample(n=len(df_majority)//2, random_state=123)

# Upsample minority class
df_minority_upsampled = df_minority.sample(n=len(df_minority)*2, replace=True, random_state=123)

# Combine minority class with downsampled majority class
df_resampled = pd.concat([df_majority_downsampled, df_minority_upsampled])

# Shuffle the dataset to randomize the order of samples
df_resampled = df_resampled.sample(frac=1, random_state=123)

# Display class counts to verify balance
print(df_resampled['Exited'].value_counts())

1    2602
0    2549
Name: Exited, dtype: int64


<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment</h2>
    
Nice code! It's great that you use `sample` and define `random state`. 

</div>
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2>  Reviewer's comment </h2>
   

It would be perfect if you defined it as a method that can be called. 
    
</div>
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
Resampling should be applied to train data only. 

</div>

^ That does make more sense, now I don't have to split the data up again too

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
Exacltly. 

</div>

In [19]:
# Initialize new training variables
features_train = df_resampled.drop(['Exited'], axis=1)
target_train = df_resampled['Exited']

In [20]:
# Fit new tree model to new resampled data

'''
best_f1 = 0
best_accuracy = 0
best_est = 0
for est in range(1, 21):
    tree = DecisionTreeClassifier(random_state=12345, max_depth=est)
    tree.fit(features_train, target_train)
    # Predict on validation set to calculate F1 score
    tree_predicted_valid = tree.predict(features_valid)
    # Calculate accuracy score on validation set
    accuracy = accuracy_score(target_valid, tree_predicted_valid)
    # Calculate F1 score
    f1 = f1_score(target_valid, tree_predicted_valid)
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_est = est
    if f1 > best_f1:
        best_f1 = f1
        best_est = est

print("Accuracy of the best model on the validation set (max_depth = {}): {}".format(best_est, best_accuracy)) # 18
print("Best f1 score of the best model on the validation set (max_depth = {}): {}".format(best_est, best_f1)) # 18
'''
tree = DecisionTreeClassifier(random_state=12345, max_depth=18)
tree.fit(features_train, target_train)

tree_predicted_valid = tree.predict(features_valid)

print('Recall:', recall_score(target_valid, tree_predicted_valid))
print()
print('Precision:', precision_score(target_valid, tree_predicted_valid))
print()
print('F1 score:', f1_score(target_valid, tree_predicted_valid))
print()
print('Valid acc:', accuracy_score(target_valid, tree_predicted_valid))
print()
print('AUC-ROC value:', roc_auc_score(target_valid, tree_predicted_valid))

Recall: 0.6472491909385113

Precision: 0.4424778761061947

F1 score: 0.5256241787122208

Valid acc: 0.774375

AUC-ROC value: 0.7260258348185972


<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
Although it may be useful in some cases, using ``class_weight` here is not necessary. 
    
</div>

# Findings with balanced data (DecisionTree)

Training the tree model on the sampled data increased recall by ~.19, reduced precision by ~.31, reduced the f1 score by ~.04, lowered the accuracy by ~.08, and increased the AOC-ROC value by ~.2.

Overall a reduction in quality because the f1 score was reduced.

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
Please make the conclusions broader so that a reader does not need to scroll up and read the previous one in order to compare the models. 
    
    
    
</div>
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2 </h2>
    
Great!     
</div>

In [21]:
# Trying RandomForestClassifier

'''
# Test to find optimal n_estimators value
best_f1 = 0
best_accuracy = 0
best_est = 0
for est in range(1, 101): #
    forest = RandomForestClassifier(random_state=54321, n_estimators=est) # set number of n_estimators
    forest.fit(features_train, target_train) # train model on training set
    forest_predicted_valid = forest.predict(features_valid)
    accuracy = accuracy_score(target_valid, forest_predicted_valid) # calculate accuracy score on validation set
    f1 = f1_score(target_valid, forest_predicted_valid)
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_est = est
    if f1 > best_f1:
        best_f1 = f1
        best_est = est

print("Accuracy of the best model on the validation set (n_estimators = {}): {}".format(best_est, best_accuracy)) # 39
print("Best f1 score of the best model on the validation set (n_estimators = {}): {}".format(best_est, best_f1)) # 39
'''

forest = RandomForestClassifier(random_state=54321, n_estimators=39)
forest.fit(features_train, target_train)

forest_predicted_valid = forest.predict(features_valid)

print('Without class_weight:')
print()
print('Recall:', recall_score(target_valid, forest_predicted_valid))
print()
print('Precision:', precision_score(target_valid, forest_predicted_valid))
print()
print('F1 score:', f1_score(target_valid, forest_predicted_valid))
print()
print('Valid acc:', accuracy_score(target_valid, forest_predicted_valid))
print()
print('AUC-ROC value:', roc_auc_score(target_valid, forest_predicted_valid))

Without class_weight:

Recall: 0.6699029126213593

Precision: 0.5734072022160664

F1 score: 0.617910447761194

Valid acc: 0.84

AUC-ROC value: 0.7753077692463884


# Findings with balanced data (RandomForest)

Training the forest model on the new resampled data increased recall by almost .2, reduced precision by ~.22, increased the f1 score by ~.03, reduced the accuracy by ~.03, and increased the AOC-ROC value by ~.05.

Overall a increase in quality because the F1 score increased.

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2>   Reviewer's comment 4 ✔️</h2>
    
Very good.

</div>

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Please add AUC-ROC score. 
    
</div>

In [22]:
# Trying Logistic Regression

logReg = LogisticRegression(random_state=12345, solver='liblinear')
logReg.fit(features_train, target_train)

logReg_predicted_valid = logReg.predict(features_valid)

print('Recall:', recall_score(target_valid, logReg_predicted_valid))
print()
print('Precision:', precision_score(target_valid, logReg_predicted_valid))
print()
print('F1 score:', f1_score(target_valid, logReg_predicted_valid))
print()
print('Valid acc:', accuracy_score(target_valid, logReg_predicted_valid))
print()
print('AUC-ROC value:', roc_auc_score(target_valid, logReg_predicted_valid))

Recall: 0.6796116504854369

Precision: 0.3115727002967359

F1 score: 0.427263479145473

Valid acc: 0.648125

AUC-ROC value: 0.6601001707113475


# Findings with balanced data (LogisticReg)

Training the LogisticRegression model on the resampled data greatly increased the recall by about .66, reduced the precision by ~.36, increased the F1 score by ~.4, reduced the accuracy by ~.16, and increased the AOC-ROC value by ~.15.

Definitely a better model than it was before, but still not good.

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Please don't forget to add AUC-ROC score and try this model in the previous section, with the unbalanced data. 
    
</div>

**Random Forest Classifier performed the best**

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
Yes, it indeed performs better. Now we can evaluate F1 score on test data. 
    
</div>

In [23]:
# Perform final test with Random Forest Classifier on test df

forest_predicted_test = forest.predict(features_test)

print('Recall:', recall_score(target_test, forest_predicted_test))
print()
print('Precision:', precision_score(target_test, forest_predicted_test))
print()
print('F1 score:', f1_score(target_test, forest_predicted_test))
print()
print('Valid acc:', accuracy_score(target_test, forest_predicted_test))
print()
print('AUC-ROC value:', roc_auc_score(target_test, forest_predicted_test))

Recall: 0.6276346604215457

Precision: 0.5763440860215053

F1 score: 0.600896860986547

Valid acc: 0.822

AUC-ROC value: 0.7511981312279374


<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2>  Reviewer's comment 4 </h2>
   

By the way, on final tests of your model, you can train it on full data (train + valid) and then check the score on test data.
    
</div>

# Overview

- Prepared the data for training
- Provided explanation of preprocessing steps
- Investigated the balance of classes within the dataset
- Studied the model without considering class imbalance
- Split the data into training, validation, and testing sets
- Implemented techniques to address class imbalance
- Ensured proper model training, validation, and final testing procedures were followed
- Achieved a high F1 score
- Examined AUC-ROC values to assess model performance

# Conclusion on Class Weights

After experimenting with class weights in the models, it became evident that their impact varied across different classifiers. Notably, the use of balanced class weights generally resulted in either minor improvements or slight deteriorations in model performance. 

### DecisionTreeClassifier with class_weight

Introducing a balanced class weight to the DecisionTree model led to a decrease in accuracy by approximately 0.06, while increasing recall by around 0.1. However, this adjustment also resulted in a reduction in the F1 score by approximately 0.04 and a significant decrease in precision by about 0.26. Considering the negligible change in the AUC-ROC, it's concluded that the model without class_weight performs better.

### RandomForestClassifier with class_weight

Applying a balanced class weight to the RandomForest model maintained the recall at the same level as the unbalanced model. However, it significantly decreased precision by approximately 0.5, leading to a minor reduction in the F1 score by around 0.01 and a notable decrease in accuracy by approximately 0.1. Additionally, the AUC-ROC decreased slightly by about 0.004. Thus, it's determined that the model without a balanced class weight performs better.

### LogisticRegression with class_weight

Balancing the class weight for the LogisticRegression model notably improved recall by approximately 0.7, although it also led to a substantial reduction in precision by around 0.31. This improvement resulted in a significant increase in the F1 score by approximately 0.45. However, there was a considerable decrease in accuracy by about 0.11, alongside an increase in the AUC-ROC by approximately 0.19. Despite these improvements, the overall performance remains unsatisfactory.

In conclusion, while the impact of class_weight='balanced' varied across models, it generally did not significantly enhance performance. Moreover, it's evident that LogisticRegression consistently underperforms compared to DecisionTree and RandomForest models, especially before balancing the classes.


# Conclusion

After resampling the data, significant effects were observed on the models. The RandomForest model demonstrated the most notable improvement, particularly in its F1 score, which experienced a substantial increase due to achieving similar recall and precision values (approximately 0.62). Conversely, the DecisionTree model showcased a peculiar response to the resampled data, as it encountered a considerable decrease in precision despite maintaining relatively unchanged performance in other metrics.

Among the selected models, LogisticRegression displayed the poorest performance. Despite minimal changes observed with the balanced data, it exhibited a slight deterioration in all aspects, further emphasizing its inadequate performance even before the introduction of new data.

When examining each model's performance with balanced data individually:

- The DecisionTree model saw a notable increase in recall (~0.19) but suffered a significant decrease in precision (~0.31), resulting in a slight reduction in the F1 score (~0.04).
- The RandomForest model experienced a substantial increase in recall (~0.2) and a slight improvement in the F1 score (~0.03), despite a decrease in precision (~0.22).
- The LogisticRegression model exhibited a remarkable boost in recall (~0.66) but also a significant reduction in precision (~0.36), resulting in a substantial increase in the F1 score (~0.4).

While each model responded differently to the resampled data, it is evident that careful consideration is necessary when interpreting their performances in the context of class imbalance and data preprocessing.


<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Please elaborate on it. What can we infer? In the text above, you provided the results (what has been done), but can we conclude from these results? Did the resampling help? Which model shows the best performance? What can be improved in these models? 
    
</div>
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 2</b>
    

Looks very good! Please don't forget to update the conclusions after the updates. 
    
</div>
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 3</b>
    

Let's keep this comment relevant because updates in the OHE method may cause some changes.     
</div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 4</h2>
    
Great job, thank you so much!     
</div>